features/323-login-id/design.md

```diff
@@ -0,0 +1,185 @@
+- Realm is a user specified value, auth gear does not concern its semantics.
+- Default realm is 'default' if unspecified, for application without need of compound login ID.
+- Possible realm values are whitelisted.
+- Realm would be considered when matching login ID with principal.
+  i.e. password authenticated principals has unique (realm, login ID).
+- API consumer could specify realm along with login ID when realm is required.
+
+This solution should support majority of use case for compound login ID.
+
+For example, when logging in with realms:
+```json
+{ "realm": "teacher", "login_id": "a@example.com", "password": "12345678" }
+{ "realm": "student", "login_id": "a@example.com", "password": "12345678" }
+```
+If authentication succeed, both request would result in same user, with
+different principals.
+
+This addresses use case 10.
+
+## Allow Multiple Login ID per Login ID Key
+In some website, user can associate multiple emails (potentially unlimited) with
+same user. These email may be used in auth features:
+- Login with multiple emails
+- Backup email
+- Welcome email
+
+A login ID key should allow having multiple associated login ID.
+When signing up, user should provide a list of login ID. e.g.
+```json
+{
+    "login_ids": [
+        { "username": "kiootic" },
+        { "email": "kiootic+1@oursky.com" },
+        { "email": "kiootic+2@oursky.com" }
+    ],
+    "password": "12345678"
+}
+```
+
+This addresses use case 6.
+
+
+## Standard Keys
+With reference to OpenID Connect standard claims, a list of standard keys
+would be specified:
+- `username`
+- `email`
+- `phone`
+- etc.
+
+These keys would be used in user metadata.
+
+
+## Configurable Login ID Keys
+The login ID keys are whitelisted by developer. In additional, developer can
+configure each login ID key:
+- The corresponding standard key (normalization may be applied)
+- The minimum/maximum amount of login ID
+
+The auth gear would use configured keys in password recovery and welcome message.
+
+
+## User Settings
+User settings would be added. This settings is a map of strings, and only keys
+recognized by auth gear is allowed:
+- `primary_login_id`
+
+The settings have default values, and its value can be updated later using API.
+
+This settings is intended to be consumed by auth features:
+- Welcome email/SMS: Can be configured to send to primary login ID/all supported login ID.
+
+# Design Detail
+This specification would be implemented in stages:
+1. [Implicit login ID key, Login ID realm](stage-1.md)
+2. [Multiple login IDs per key, Configurable login ID keys, Standard keys](stage-2.md)
+3. [User settings](stage-3.md)
+
+Design detail of each stage can be found in corresponding documents.
+
+# Appendix
+
+**OpenID Connect: Standard Claims**: https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+
+## Example user object
+```diff
+    {
+        "user_id": "5bf1e4d2-e1c4-4517-93c7-7dae89261da6",
+-       "login_ids": [
```

Comment:
So we don't support multiple login IDs any more after this change?

Reply:
We do support multiple login ID. However, the list will not be shown in authn responses. Instead, it should be included as part of the principal APIs later.

Reply:
What is principal APIs?

Reply:
I'll open issues for this later, along with some other potential issues discovered during the design.

Reply:
As discussed, please make your case on why a principal APIs are needed. And why it is better.

A reminder is there was an effort to try to unify password authentication / SSO / and other auth methods before, but the result is too complicated and no one really uses that API, so turns out we fallback to make it simple to use PW/SSO (and passcode was proposed in #294) in separate APIs; So the principals APIs need to take care of these too. 

I'm just guessing as I still don't know what is the scope of principals API.

Reply:
On authn responses, shall we list all loginIDs actually? If not, I think at least we need to have the login ID used?

Reply:
This is another thing we need to discuss in principals APIs (sorry for deferring many things to it): if we had `primary_login_id`, do we have `primary_oauth_id`, `primary_custom_token_id`, etc.? It's better to unify as a single `primary_principal_id` (or whatever name we use).

Also, I thought about it in more details and seems we do not need this field at the moment. I'll discuss with @louischan-oursky and see if we can just remove this field.

Reply:
Actually, my question will be what is `primary_login_id` (or `primary_principal_id`), how would it be defined OR what would it be used for?

If it is configured as the "primary email to send welcome email/forgot password etc to", or "primary SMS number to send login code to", it seems make more sense to have a primary metatag to the login ID INSIDE each specific keys.

(so I still don't see why we need `principal` or `principal_id` at all, please don't defer anything to something we probably won't have / can't make the justification to have it yet)

Reply:
The use of `primary_login_id` is specified in stage 3: It is used to determine the destination of welcome email (all email login ID/primary email login ID).

Reply:
So except primary email, we might also have primary SMS number, so we don't have one single primary login ID, but primary login ID for each login ID Key type as described above?

Reply:
I didn't consider the multiple primary login ID for email & SMS case. I'll think about it.